rust/arrow/src/csv/writer.rs

```diff
@@ -0,0 +1,363 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! CSV Writer
+//!
+//! This CSV writer allows Arrow data (in record batches) to be written as CSV files.
+//! The writer does not support writing `ListArray` and `StructArray`.
+//!
+//! Example:
+//!
+//! ```
+//! use arrow::array::*;
+//! use arrow::csv;
+//! use arrow::datatypes::*;
+//! use arrow::record_batch::RecordBatch;
+//! use std::fs::File;
+//! use std::sync::Arc;
+//!
+//! let schema = Schema::new(vec![
+//!     Field::new("c1", DataType::Utf8, false),
+//!     Field::new("c2", DataType::Float64, true),
+//!     Field::new("c3", DataType::UInt32, false),
+//!     Field::new("c3", DataType::Boolean, true),
+//! ]);
+//!
+//! let c1 = BinaryArray::from(vec![
+//!     "Lorem ipsum dolor sit amet",
+//!     "consectetur adipiscing elit",
+//!     "sed do eiusmod tempor",
+//! ]);
+//! let c2 = PrimitiveArray::<Float64Type>::from(vec![
+//!     Some(123.564532),
+//!     None,
+//!     Some(-556132.25),
+//! ]);
+//! let c3 = PrimitiveArray::<UInt32Type>::from(vec![3, 2, 1]);
+//! let c4 = PrimitiveArray::<BooleanType>::from(vec![Some(true), Some(false), None]);
+//!
+//! let batch = RecordBatch::new(
+//!     Arc::new(schema),
+//!     vec![Arc::new(c1), Arc::new(c2), Arc::new(c3), Arc::new(c4)],
+//! );
+//!
+//! let file = File::create("/tmp/out.csv").unwrap();
+//!
+//! let writer = csv::Writer::new(file);
+//! writer.write(vec![&batch, &batch]).unwrap();
+//! ```
+
+use csv as csv_crate;
+
+use std::fs::File;
+
+use crate::array::*;
+use crate::datatypes::*;
+use crate::error::{ArrowError, Result};
+use crate::record_batch::RecordBatch;
+
+fn write_primitive_value<T>(array: &ArrayRef, i: usize) -> String
+where
+    T: ArrowNumericType,
+    T::Native: ::std::string::ToString,
+{
+    let c = array.as_any().downcast_ref::<PrimitiveArray<T>>().unwrap();
+    c.value(i).to_string()
+}
+
+/// A CSV writer
+pub struct Writer {
+    /// The file to write to
+    file: File,
```

Comment:
It may be a better idea to accept a Box<Write> as target. This can make your implementation more flexible.

Reply:
We can refactor this with a separate PR, that ideally does the same to the CSV reader, so we're consistent with our IO APIs

Reply:
Great, looking forward to see the following PR